src/v/kafka/client/consumer.cc

```diff
@@ -475,22 +478,133 @@ ss::future<fetch_response> consumer::fetch(
       detail::reduce_fetch_response);
 }
 
+template<typename request_factory>
+ss::future<
+  typename std::invoke_result_t<request_factory>::api_type::response_type>
+consumer::reset_coordinator_and_retry_request(request_factory req) {
+    return find_coordinator_with_retry_and_mitigation(
+             _gate,
+             _config,
+             _brokers,
+             group_id(),
+             name(),
+             [this](std::exception_ptr ex) { return _external_mitigate(ex); })
+      .then(
+        [this, req{std::move(req)}](shared_broker_t new_coordinator) mutable {
+            _coordinator = new_coordinator;
+            // Calling req_res here will re-issue the request on the
+            // new coordinator
+            return req_res(std::move(req));
+        });
+}
+
+template<typename request_factory, typename response_t>
+ss::future<response_t>
+consumer::maybe_process_response_errors(request_factory req, response_t res) {
+    auto me = shared_from_this();
+    // By default, look at the top-level for errors
+    switch (res.data.error_code) {
+    case error_code::not_coordinator:
+        vlog(
+          kclog.debug,
+          "Wrong coordinator on consumer {}, getting new coordinator "
+          "before retry",
+          *me);
+        return reset_coordinator_and_retry_request(std::move(req));
+    default:
+        // Return the whole response otherwise
+        return ss::make_ready_future<response_t>(std::move(res));
+    }
+}
+
+template<typename request_factory>
+ss::future<metadata_response> consumer::maybe_process_response_errors(
+  request_factory req, metadata_response res) {
+    auto me = shared_from_this();
+    for (auto& topic : res.data.topics) {
+        switch (topic.error_code) {
+        case error_code::not_coordinator:
+            vlog(
+              kclog.debug,
+              "Wrong coordinator on consumer {}, topic {}, getting new "
+              "coordinator before retry",
+              *me,
```

Comment:
why is a shared pointer to `this` being created? it looks like you can replace `*me` in the vlog statement with `this` and remove `me` entirely?

Reply:
No reason, I use it as an "alias" for `this` (or `shared_from_this`) because it is more readable IMO.

Reply:
i would avoid using shared pointers for this. it makes the code look like lifetime issues are important in the code (shared pointers implement reference counted pointers), but it doesn't appear that life time is important.

if you want readability you could do `auto me = this`.